srtcore/api.cpp

```diff
@@ -2333,99 +2351,48 @@ int CUDT::getsockopt(
 
 int CUDT::setsockopt(SRTSOCKET u, int, SRT_SOCKOPT optname, const void* optval, int optlen)
 {
-   try
-   {
-      CUDT* udt = s_UDTUnited.locateSocket(u, s_UDTUnited.ERH_THROW)->m_pUDT;
-      udt->setOpt(optname, optval, optlen);
-      return 0;
-   }
-   catch (const CUDTException& e)
+   if (!optval)
    {
-      s_UDTUnited.setError(new CUDTException(e));
-      return ERROR;
+       return setError(MJ_NOTSUP, MN_INVAL, 0);
    }
```

Comment:
So this is to collapse similar places, but the intent is now lost. `CUDT::setError` calls `s_UDTUnited.setError(...)` and always returns `SRT_ERROR`, independent on the result of `s_UDTUnited.setError(...)`. Is it going to be reused in the upcoming changes?
```
int CUDT::setError(CodeMajor mj, CodeMinor mn, int syserr)
{
    s_UDTUnited.setError(new CUDTException(mj, mn, syserr));
    return SRT_ERROR;
}
```

Reply:
Yes. That's just for simplification. The old-new comparison here is misleading - this line, together with three new, is completely new here. Other places are intended to be changed, I just use this much more in the new code. the `CUDTUnited::setError` is another story - there's a doubt as to whether the `new` operator is the best idea to be called in the handler of the memory allocation error. There's another PR raised for this.

Reply:
The point here is that when calling a function named `setError(..)`, I expect it will either succeed or fail. And the return statement should provide the result of executing the function.
But here `SRT_ERROR` is always returned, and it has no connection to calling
`s_UDTUnited.setError(new CUDTException(mj, mn, syserr));`

Although I am not a fan of macros, but macros looks more natural here:
```
#define SET_ERROR_AND_RET(MJ, MN, SYSERR) \
} \
    s_UDTUnited.setError(new CUDTException(mj, mn, syserr)); \
    return SRT_ERROR; \
}
```

Or maybe a better name should be considered.

Reply:
This is why this function is called `setError`. If it was intended to return anything else than SRT_ERROR, it would be called `checkForError` or something like that. Here the intention is only one thing: we know we have error, we need to define which error it is and leave the trace at the places where it is predicted. The return value itself is actually not so important here, most important is the error record that is written into the thread-specific data - at best the return value is only a part of the error information.

It's made to return this value so that you could do
```
return setError(ERROR_SPEC);
```
instead of necessarilessly longer
```
setError(ERROR_SPEC);
return SRT_ERROR;
```
which makes sense moreover that this whole instruction is intended to serve exactly ONE error report, while two separate instructions suggest there are two different things done.

Maybe a macro would bind them too, but I don't understand what the point is of hiding the fact of using the `return` instruction, while this is one of the things that should be clearly visible.

I'd better add the `[[nodiscard]]` attribute to this `setError` function so that the compiler can check it is always used with return (though won't prevent the code where the user saved it in the variable, but this sounds too stupid).

Reply:
I fully agree with Maxim: having a `set()` method return a value unrelated to the set operation is confusing, is a deviation from long-established norms, and is masking the intent. Someone looking at that code would have to remember your intent of the day by heart or, otherwise, will have no choice to go and check what the function does or if anything changed in it that would affect the caller's return. And what it does, is also semantically wrong: for example `setError(NOERROR)` would for some reason return ERROR. The common convention is `void set(type value)` unless the *set function itself* can fail and that's exactly how it is defined underneath the wrappers: `void CUDTUnited::setError(CUDTException* e) {}`. `CUDT::setError(...)` wrappers should be changed to match that because there is nothing in them that can actually fail or needs to be communicated back to the caller.

Reply:
I understand the reasoning about the name, and I can agree to any proposal to change it. Still, the intention behind this construction is that:
1. This is needed to be argument for return, and it should be clear that it will return always the error value.
2. This is a statement used exclusively in situations when error is expected and it should be nothing else expected as a return value but this system error value -1. Simply usage of this function should be clear enough that the only possible return value is -1, always.
3. This statement - together with the `[[nodiscard]]` modifier - should also ensure that setting this error is always connected with immediate returning with -1.

Could be the name changed to something like `errorWithType(...)` or something of that kind. But replacement with a void-returning setting and return following with explicit -1 value is making the code worse and more error-prone.

Reply:
There is no need for any of it. The return of the setError() has nothing to do with the return of the caller, there is absolutely no need to bind the two, it is confusing. There is no need to save that extra line, - having clear code in two lines is better than confusing shortcuts done in one. Please leave setError() as void unless the function itself can fail (and today it cannot).